### PR TITLE
[API-1219] Add zero-config support for Java records to compact serialization

### LINKED ISSUE #19753: Javas 16 Records support in Compact Serializer [API-1219]
Add support for Java 16 record classes in Compact Serialization for no-configuration use case.

Support should be added for classes in top level and as fields.

### PR DESCRIPTION
**Pull request by mdumandag**
With the recent Java versions, records provide a convenient way
of defining immutable data classes.

We do support zero-config use cases with the reflection for regular
Java classes, but records do require special handling as our
current way of setting individual fields one-by-one does not work
with them.

Instead, we have to use the canonical constructor of the record
class. That requires us to pass the fields to this constructor
in the expected order.

The flow of using the new Java record serializer is as follows:

- When we cannot find a serializer for a type, as a last
resort, we try to use compact serialization.
- While setting up the reflective serializer for that object,
we check if the class of that type is a record or not, with
the https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Class.html#isRecord().
- If it is a record, we use a newly introduced java record serializer.
- If not, we use the reflective serializer we already have.

The record serializer extracts the component readers/writers out
of the record class, in the same order as they are defined
in the canonical constructor.

The readers support using default values of the types when such field
does not exist, so that we can have a frictionless schema evolution
experience.

The types of fields(components) we support are the same with the
reflective serializer.

Closes #19753

### CODE REVIEW

**Comment by mdumandag** [comment]
run-with-jdk17

**Comment by mdumandag** [comment]
run-with-jdk17

**Comment by mdumandag** [comment]
run-with-jdk17

**Review by ihsandemir — APPROVED** [review]

File: hazelcast-it/jdk17-tests/src/test/java/com/hazelcast/serialization/compact/record/AllTypesRecord.java [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+                null,
+                null,
+                null,
+                null,
+                null
+        );
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply hazelcast-it/jdk17-tests/src/test/java/com/hazelcast/serialization/compact/record/AllTypesRecord.java:184]: For arrays. The provided equals method checks for object identity for arrays, but we need to compare the contents

**Review by sancar — APPROVED** [review]

**Pull request merged**